fuzz: wallet: add target for `MigrateToDescriptor`

brunoerg opened the pull request:
This PR adds fuzz coverage for the scriptpubkeyman migration (`MigrateToDescriptor`). Note that it's a test for the migration of the scriptpubkey manager, not for the whole migration process as tried in #29694, because:
1) The wallet migration deals with DBs which is expensive for fuzzing (was getting around 3 exec/s);
2) Mocking would require lots of refactors.

This target loads keys, HDChain (even inactive ones), watch only and might add tons of different scripts, then calls `MigrateToDescriptor`. It does not play with encrypted stuff because it would make the target super slow. Also, after the migration there are some assertions that would work as a regression test for https://github.com/bitcoin/bitcoin/pull/31452, for example.

brunoerg committed:
fuzz: wallet: add target for `MigrateToDescriptor`

brunoerg commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -50,6 +52,13 @@ void initialize_spkm()
+{
+    static const auto testing_setup{MakeNoLogFileContext<const TestingSetup>()};
+    g_setup = testing_setup.get();
+    SelectParams(ChainType::MAIN);
```
Just removed it. Redudant.

brunoerg commented:
08b574dde2c5f13a4c9dadcb9d8bded158c59623..5f2fb4dd333f8d1db27c78414afc84454e6ee4a5: Address https://github.com/bitcoin/bitcoin/pull/32624#discussion_r2109086986

w0xlt reviewed (commented):
Approach ACK

marcofleon commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -204,5 +211,137 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+        );
+    }
+
+    auto result{legacy_data.MigrateToDescriptor()};
```
No null check after this? Doesn't seem to ever return null based on the coverage report, but could still be worth having to avoid potential undefined behavior.

marcofleon commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -204,5 +211,137 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+    bool watch_only = false;
+    const auto pub_key = ConsumeDeserializable<CPubKey>(fuzzed_data_provider);
+    if (!pub_key || !pub_key->IsFullyValid()) return;
+    if (legacy_data.LoadWatchOnly(GetScriptForDestination(WitnessV0KeyHash{*pub_key}))) watch_only = true;
```
Is it worth testing other script types here as well? Not super familiar with the wallet code, so disregard if irrelevant.

marcofleon reviewed (commented):
Left some questions below.

Also, [coverage report](https://marcofleon.github.io/coverage/spkm_migration/) after about a day of fuzzing.

brunoerg commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -204,5 +211,137 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+    }
+
+    auto result{legacy_data.MigrateToDescriptor()};
+    auto hd_chain_version{legacy_data.GetHDChain().nVersion};
```
Nice catch, we can simply check `hd_chain.nVersion` instead of having `auto hd_chain_version{legacy_data.GetHDChain().nVersion};`. We could also add an assert to ensure these two values are the same.

brunoerg commented:
Force-pushed [6436ad7](https://github.com/bitcoin/bitcoin/commit/6436ad7f2f93774cb719c89965800b51f546420c) to [6227441](https://github.com/bitcoin/bitcoin/commit/6227441b782b50403571d3d97e8aa174c95fc511):

- Added `PKHash` for `LoadWatchOnly`
- Addressed https://github.com/bitcoin/bitcoin/pull/32624#discussion_r2521415641
- Addressed https://github.com/bitcoin/bitcoin/pull/32624#discussion_r2521400460

marcofleon commented:
Nice, lgtm ACK 51917d94e657024cc0af8aa748a9d5d39a617e85

frankomosh commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -204,5 +211,142 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+    auto result{legacy_data.MigrateToDescriptor()};
+    assert(result);
+    size_t added_chains{static_cast<size_t>(add_hd_chain) + static_cast<size_t>(add_inactive_hd_chain)};
+    if ((add_hd_chain && version >= CHDChain::VERSION_HD_CHAIN_SPLIT) || (!add_hd_chain && add_inactive_hd_chain)) {
```
There might be a small issue with how `version` is used in the assertion here. The `version` variable is initialized at line 244 and only assigned to the active HD chain when `add_hd_chain` is true (line 249). However, when `add_inactive_hd_chain` is true, the inactive chain gets its own version choice at line 257.

Challenge arises when `add_hd_chain` is false but `add_inactive_hd_chain` is true. In this scenario, condition `(!add_hd_chain && add_inactive_hd_chain)` would trigger `added_chains *= 2`, but this doesn't account for whether the `inactive` chain's version is actually VERSION_HD_CHAIN_SPLIT or not , as it just assumes any inactive 
chain should double the count.

If you consider that these sequence of events might cause assertion miscalculations then maybe a solution might be store and checking the inactive chain's version separately or checking `hd_chain.nVersion` directly after it's set.

brunoerg replied:
> Challenge arises when add_hd_chain is false but add_inactive_hd_chain is true. In this scenario, condition (!add_hd_chain && add_inactive_hd_chain) would trigger added_chains *= 2, but this doesn't account for whether the inactive chain's version is actually VERSION_HD_CHAIN_SPLIT or not , as it just assumes any inactive chain should double the count.

We just need to check `VERSION_HD_CHAIN_SPLIT` for the active one because it's how we know if it can support hd split feature. 

```cpp
// Handle HD keys by using the CHDChains
std::vector<CHDChain> chains;
chains.push_back(m_hd_chain);
for (const auto& chain_pair : m_inactive_hd_chains) {
    chains.push_back(chain_pair.second);
}

bool can_support_hd_split_feature = m_hd_chain.nVersion >= CHDChain::VERSION_HD_CHAIN_SPLIT;
```

frankomosh reviewed (commented):
Left a few observations and questions below.

brunoerg commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -204,5 +211,142 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+    if (legacy_data.LoadWatchOnly(script_dest)) watch_only = true;
+
+    size_t added_script{0};
+    bool good_data{true};
```
Yes, see https://github.com/bitcoin/bitcoin/pull/28815.

frankomosh replied:
Did not know that. Thanks for the context.

frankomosh reviewed (commented):
Code Review ACK 779e7825dbfba6a1ebd6ad62a8b2f312bd2c6b5f

marcofleon reviewed (commented):
reACK 779e7825dbfba6a1ebd6ad62a8b2f312bd2c6b5f

Just minor rebases since last review.
